File: ios/Classes/PostPurchase/Method/KlarnaPostPurchaseSDKMethod.swift
import Flutter

protocol KlarnaPostPurchaseSDKMethod {
    var id: Int { get }
}

class KlarnaPostPurchaseSDKMethods {
    
    struct Create: KlarnaPostPurchaseSDKMethod, Decodable {
        let id: Int
        let environment: String?
        let region: String?
        let resourceEndpoint: String?
    }

    struct Initialize: KlarnaPostPurchaseSDKMethod, Decodable {
        let id: Int
        let locale: String
        let purchaseCountry: String
        let design: String?
    }
    
    struct AuthorizationRequest: KlarnaPostPurchaseSDKMethod, Decodable {
        let id: Int
        let clientId: String
        let scope: String
        let redirectUri: String
        let locale: String?
        let state: String?
        let loginHint: String?
        let responseType: String?
    }
    
    struct RenderOperation: KlarnaPostPurchaseSDKMethod, Decodable {
        let id: Int
        let operationToken: String
        let locale: String?
        let redirectUri: String?
    }
    
    struct Destroy: KlarnaPostPurchaseSDKMethod {
        let id: Int
    }

    class Parser: MethodParser<KlarnaPostPurchaseSDKMethod> {
        override func parse(call: FlutterMethodCall) throws -> KlarnaPostPurchaseSDKMethod? {
            switch call.method {
            case "create":
                return call.decode(Create.self)
            case "initialize":
                return call.decode(Initialize.self)
            case "renderOperation":
                return call.decode(RenderOperation.self)
            case "authorizationRequest":
                return call.decode(AuthorizationRequest.self)
            case "destroy":
                return Destroy(id: try call.requireArgument(key: "id"))
            default:
                return nil
            }
        }
    }
}

File: android/src/main/kotlin/com/klarna/mobile/sdk/flutter/postpurchase/method/KlarnaPostPurchaseSDKMethod.kt
package com.klarna.mobile.sdk.flutter.postpurchase.method

import com.klarna.mobile.sdk.flutter.core.method.MethodParser
import com.klarna.mobile.sdk.flutter.core.util.requireArgument
import io.flutter.plugin.common.MethodCall

internal sealed class KlarnaPostPurchaseSDKMethod(val id: Int) {
    class Create(
        id: Int,
        val environment: String?,
        val region: String?,
        val resourceEndpoint: String?
    ) : KlarnaPostPurchaseSDKMethod(id)

    class Initialize(
        id: Int,
        val locale: String,
        val purchaseCountry: String,
        val design: String?
    ) : KlarnaPostPurchaseSDKMethod(id)

    class AuthorizationRequest(
        id: Int,
        val clientId: String,
        val scope: String,
        val redirectUri: String,
        val locale: String?,
        val state: String?,
        val loginHint: String?,
        val responseType: String?
    ) : KlarnaPostPurchaseSDKMethod(id)

    class RenderOperation(
        id: Int,
        val operationToken: String,
        val locale: String?,
        val redirectUri: String?
    ) :
        KlarnaPostPurchaseSDKMethod(id)

    class Destroy(id: Int) : KlarnaPostPurchaseSDKMethod(id)

    internal object Parser : MethodParser<KlarnaPostPurchaseSDKMethod> {
        override fun parse(call: MethodCall): KlarnaPostPurchaseSDKMethod? {
            return when (call.method) {
                "create" -> Create(
                    call.requireArgument("id"),
                    call.argument("environment"),
                    call.argument("region"),
                    call.argument("resourceEndpoint")
                )
                "initialize" -> Initialize(
                    call.requireArgument("id"),
                    call.requireArgument("locale"),
                    call.requireArgument("purchaseCountry"),
                    call.argument("design")
                )
                "authorizationRequest" -> AuthorizationRequest(
                    call.requireArgument("id"),
                    call.requireArgument("clientId"),
                    call.requireArgument("scope"),
                    call.requireArgument("redirectUri"),
                    call.argument("locale"),
                    call.argument("state"),
                    call.argument("loginHint"),
                    call.argument("responseType")
                )
                "renderOperation" -> RenderOperation(
                    call.requireArgument("id"),
                    call.requireArgument("operationToken"),
                    call.argument("locale"),
                    call.argument("redirectUri")
                )
                "destroy" -> Destroy(
                    call.requireArgument("id")
                )
                else -> null
            }
        }

    }
}
